src/Features/Core/Portable/EditAndContinue/DebuggingSession.cs

```diff
@@ -1,20 +1,327 @@
+        internal ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>> NonRemappableRegions { get; private set; }
+
+        /// <summary>
+        /// All module updates that have been committed or deferred.
+        /// When a module whose MVID is in this map is loaded while the debugee is running all updates that are listed here are applied.
+        /// </summary>
+        /// <remarks>
+        /// Note that a delta needs to be applied to each loaded module instance.
+        /// The first application must apply both symbol and metadata deltas, the next application must only apply metadata.
+        /// This is because the debugger shares the SymReader for all module instances.
+        /// TODO: ideally this implementation detail would not leak through and this would be handled by the debugger.
+        /// </remarks>
+        private readonly Dictionary<Guid, List<Deltas>> _moduleUpdates;
+        private readonly object _moduleUpdatesGuard = new object();
+
+        private readonly HashSet<Guid> _modulesPreparedForUpdate;
+        private readonly object _modulesPreparedForUpdateGuard = new object();
+
+        /// <summary>
+        /// The solution captured when the debugging session entered run mode (application debugging started),
+        /// or the solution which the last changes commited to the debuggee at the end of edit session were calculated from.
+        /// The solution reflecting the current state of the modules loaded in the debugee.
+        /// </summary>
+        internal Solution LastCommittedSolution { get; private set; }
+
+        internal DebuggingSession(
+            Workspace workspace,
+            IDebuggeeModuleMetadataProvider debugeeModuleMetadataProvider,
+            IActiveStatementProvider activeStatementProvider)
+        {
+            Debug.Assert(workspace != null);
+            Debug.Assert(debugeeModuleMetadataProvider != null);
+
+            DebugeeModuleMetadataProvider = debugeeModuleMetadataProvider;
+            _projectModuleIds = new Dictionary<ProjectId, (Guid, Diagnostic)>();
+            _projectEmitBaselines = new Dictionary<ProjectId, EmitBaseline>();
+            _moduleUpdates = new Dictionary<Guid, List<Deltas>>();
+            _modulesPreparedForUpdate = new HashSet<Guid>();
+
+            ActiveStatementProvider = activeStatementProvider;
+
+            LastCommittedSolution = workspace.CurrentSolution;
+            NonRemappableRegions = ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>>.Empty;
+        }
+
+        // test only
+        internal void Test_SetNonRemappableRegions(ImmutableDictionary<ActiveMethodId, ImmutableArray<NonRemappableRegion>> nonRemappableRegions)
+        {
+            NonRemappableRegions = nonRemappableRegions;
+        }
+
+        public void Dispose()
+        {
+            lock (_projectEmitBaselinesGuard)
+            {
+                _lazyBaselineModuleReaders?.ForEach(m => m.Dispose());
```

Comment:
💡 Consider moving the values from the field to a temporary local  inside the lock, and then calling `Dispose` outside it.

Reply:
Good catch.

---
In reply to: [270628378](https://github.com/dotnet/roslyn/pull/34027#discussion_r270628378) [](ancestors = 270628378)